File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.cc
#include <iostream>
#include <algorithm>
using namespace std;

int a[1010], n, m, l;

bool decision(int cur){
    int cnt = 0;
    for(int i = 1;i <= n + 1;i++){
        int diff = a[i] - a[i - 1];
        cnt += diff / cur;
        if(diff % cur == 0) cnt--;
    }

    return cnt <= m;
}

int minimazation(){
    // 정답의 범위는 1 ~ l
    int lo = 1, hi = l;
    while(lo < hi){
        int mid = (lo + hi) / 2; // 중간값
        // 결정 문제의 답이 1 이면 
        // 정답은 [lo, mid]에 존재
        if(decision(mid)) hi = mid;
        
        // 결정 문제의 답이 0 이면
        // 정답은 [mid + 1, hi]에 존재
        else lo = mid + 1;
    }

    return lo;
}

int main(){
    cin >> n >> m >> l;
    for(int i = 1;i <= n;i++) cin >> a[i];
    a[0] = 0; a[n + 1] = l;
    sort(a, a + n + 2);

    cout << minimazation();

    return 0;
}

File: 백준/Gold/1477. 휴게소 세우기/휴게소 세우기.py
import sys
input = sys.stdin.readline

n, m, l = list(map(int, input().rstrip().split()))
a = list(map(int, input().rstrip().split()))
a.append(0)
a.append(l)
a.sort()

def decision(cur):
    cnt = 0
    for i in range(1, n + 2):
        diff = a[i] - a[i - 1]
        cnt += diff // cur
        if diff % cur == 0:
            cnt -= 1
    return cnt <= m

def minimazation():
    lo = 1
    hi = l
    while(lo < hi):
        mid = (lo + hi) // 2
        if(decision(mid)):
            hi = mid
        else:
            lo = mid + 1
    return lo

print(minimazation())